File: frontend/src/mail/syncSettings.ts
// Thin wrapper over the Wails SyncSettings service. Calls are no-ops (resolved
// gracefully) when running outside the desktop app, e.g. the browser preview,
// so callers don't need to branch on the runtime.
import { SyncSettings } from '../bindings/github.com/atterpac/pigeon/internal/desktop'
import { Prefs as NotifyPrefs } from '../bindings/github.com/atterpac/pigeon/internal/desktop/notify/models'

export type NotifyPrefsInput = {
  mode: string
  mutedSenders: string[]
  quietHours: { enabled: boolean; from: string; to: string }
}

// applyNotifyPrefs pushes the user's notification preferences to the backend's
// new-mail handler. No-op (resolved gracefully) outside the desktop runtime.
export async function applyNotifyPrefs(prefs: NotifyPrefsInput): Promise<void> {
  try {
    await SyncSettings.SetNotifyPrefs(new NotifyPrefs({
      Mode: prefs.mode,
      MutedSenders: [...prefs.mutedSenders],
      QuietEnabled: prefs.quietHours.enabled,
      QuietFrom: prefs.quietHours.from,
      QuietTo: prefs.quietHours.to,
    }))
  } catch (err) {
    console.warn('applyNotifyPrefs failed', err)
  }
}

// applyPollInterval pushes the background mail-poll interval (seconds) to the
// backend, which restarts the sync loops. The backend clamps to its minimum.
export async function applyPollInterval(seconds: number): Promise<void> {
  try {
    await SyncSettings.SetPollInterval(Math.round(seconds))
  } catch (err) {
    // Non-fatal: likely no Wails runtime (browser preview) or no accounts yet.
    console.warn('applyPollInterval failed', err)
  }
}

// pollInterval reads the backend's current interval (seconds), or null when the
// runtime is unavailable.
export async function pollInterval(): Promise<number | null> {
  try {
    return await SyncSettings.PollIntervalSeconds()
  } catch {
    return null
  }
}

File: internal/desktop/syncsettings.go
package desktop

import (
	"context"

	"github.com/atterpac/email/internal/desktop/notify"
)

// SyncSettings is the thin Wails service that exposes runtime sync controls to
// the frontend, keeping App's other methods off the binding surface.
type SyncSettings struct{ app *App }

// NewSyncSettings builds the service over the given app.
func NewSyncSettings(app *App) *SyncSettings { return &SyncSettings{app: app} }

// PollIntervalSeconds reports the current background poll interval, in seconds.
func (s *SyncSettings) PollIntervalSeconds() int { return s.app.PollIntervalSeconds() }

// SetPollInterval changes the background poll interval (seconds) and restarts
// the sync loops. Values below the minimum are clamped server-side.
func (s *SyncSettings) SetPollInterval(ctx context.Context, seconds int) error {
	return s.app.SetPollInterval(ctx, seconds)
}

// NotifyPrefs reports the current notification preferences.
func (s *SyncSettings) NotifyPrefs() notify.Prefs { return s.app.NotifyPrefs() }

// SetNotifyPrefs pushes the user's notification preferences to the backend; the
// new-mail handler consults them before raising a desktop notification.
func (s *SyncSettings) SetNotifyPrefs(prefs notify.Prefs) { s.app.SetNotifyPrefs(prefs) }
